Add OpenIDP Connect Form to Cluster Settings OAuth Page

jcaianirh opened the pull request:
https://jira.coreos.com/browse/CONSOLE-1285

Create form for OpenIDP Connect, and add route to the form via the Oath Cluster Settings Page. The form follows the design located here: 
https://github.com/openshift/openshift-origin-design/blob/b1bb5ed27f9da2067f91fc24d3888c1e3efa0912/web-console/futures/oauth/oauth.md

![Screen Shot 2019-03-26 at 3 29 02 PM](https://user-images.githubusercontent.com/35978579/55031327-c0f61000-4fe4-11e9-8ad2-10b2245a4184.png)

jcaianirh commented:
@bmignano The text specified for the Client Secret states: 

> This value may also be provided in an environment variable, external file, or encrypted file.

But the field is shown as required. Should this field not be marked required?

bmignano commented:
@jcaianirh It's looking great! I think we can make that Client Secret field optional. I think we also probably want to include the default text for the claims fields (shown in the design doc).

spadgett commented:
It looks like client secret is required in the API.

@jcaianirh where is that text from?

jcaianirh commented:
@spadgett The text came from @bmignano. I think she got the text here: https://docs.openshift.com/container-platform/3.7/install_config/configuring_authentication.html

spadgett commented:
OK, this has all changed in 4.0. It's no longer optional and must be provided in a secret.

https://github.com/openshift/api/blob/master/config/v1/types_oauth.go#L496-L500

spadgett commented:
Claims is an array in the API. Are we going to provide a way to add additional claims? If not, we should at least adjust the heading and text.

There needs to be some additional help text on many of these fields.

It's not obvious to me how to enter extra scopes and extra authorize parameters since these aren't simple strings.

If we need to trim things back to the essentials, we can do that since feature freeze is rapidly approaching. We can always add to this form.

cc @enj

spadgett commented:
Related API changes pending: https://github.com/openshift/api/pull/265

jcaianirh commented:
@spadgett added api changes for openid connect except for More Options. Do we want to move forward with that part of the form for openid? Added single array for claims. Removed the outdated claims explanation text. Successfully created openid connect providers. @bmignano Added default values for claims text fields.

spadgett commented on frontend/public/components/cluster-settings/openid-idp-form.tsx:
```diff
@@ -0,0 +1,320 @@
+        namespace: 'openshift-config',
+      },
+      stringData: {
+        htpasswd: this.state.clientSecretFileContent,
```
This needs to be key `clientSecret`

```suggestion
        clientSecret: this.state.clientSecretFileContent,
```

spadgett commented on frontend/public/components/cluster-settings/openid-idp-form.tsx:
```diff
@@ -0,0 +1,320 @@
+      name,
+      type: 'OpenID',
+      challenge,
+      login,
```
`challenge` and `login` have been removed

spadgett commented on frontend/public/components/cluster-settings/openid-idp-form.tsx:
```diff
@@ -0,0 +1,320 @@
+        urls: {
+          authorize,
+          token,
+          userInfo,
```
There's one issuer URL now instead of these three

spadgett commented on frontend/public/components/cluster-settings/openid-idp-form.tsx:
```diff
@@ -0,0 +1,320 @@
+        </div>
+        <div className="form-group">
+          <label className="control-label co-required" htmlFor="tag">Mapping Method</label>
+          <Dropdown dropDownClassName="dropdown--full-width" items={mappingMethods} selectedKey={mappingMethod} title={mappingMethods[mappingMethod]} onChange={this.mappingMethodChanged} />
```
We should eventually create a common component for mapping method so we don't repeat it 9 times, but I'd leave it for now

spadgett commented on frontend/public/components/cluster-settings/openid-idp-form.tsx:
```diff
@@ -0,0 +1,320 @@
+            required />
+        </div>
+        <div className="form-group">
+          <label className="control-label co-required" htmlFor="tokenURL">Token URL</label>
```
`htmlFor` doesn't match `id` here and below, although these URLs are changing anyway

jcaianirh commented:
@spadgett updated code to address comments. need to add ca.

spadgett commented on frontend/public/components/cluster-settings/openid-idp-form.tsx:
```diff
@@ -0,0 +1,272 @@
+  }
+}
+
+type AddOpenIDPPageState = {
```
```suggestion
type AddOpenIDIDPPageState = {
```

spadgett commented on frontend/public/components/cluster-settings/openid-idp-form.tsx:
```diff
@@ -0,0 +1,272 @@
+
+type AddOpenIDPPageState = {
+  name: string;
+  mappingMethod: string;
```
This could be a string literal

```suggestion
  mappingMethod: 'claim' | 'lookup' | 'add';
```

We might want to put some IDP types into frontend/public/module/k8s/index.ts (like mapping method) to share across the forms.

spadgett commented on frontend/public/components/cluster-settings/openid-idp-form.tsx:
```diff
@@ -0,0 +1,272 @@
+  issuer: string;
+  caFileContent: string;
+  extraScopes: string;
+  extraAuthorizeParameters: string;
```
I'd suggest removing `extraScopes` and `extraAuthorizeParameters` for now. They are not simple strings.

spadgett commented on frontend/public/components/cluster-settings/openid-idp-form.tsx:
```diff
@@ -0,0 +1,272 @@
+            required />
+        </div>
+        <div className="co-form-section__separator"></div>
+        <label className="control-label">Claims</label>
```
Semantically, this really should be a header element

spadgett commented on frontend/public/components/cluster-settings/openid-idp-form.tsx:
```diff
@@ -0,0 +1,272 @@
+        </div>
+        <div className="co-form-section__separator"></div>
+        <label className="control-label">Claims</label>
+        <p className="co-help-text">The first non-empty claim is used. At least one claim is required.</p>
```
This text doesn't make sense if we only let you add one

jcaianirh replied:
@spadgett based on this text, I would think these should be required fields, right?

spadgett commented on frontend/public/components/cluster-settings/openid-idp-form.tsx:
```diff
@@ -0,0 +1,272 @@
+      apiVersion: 'v1',
+      kind: 'Secret',
+      metadata: {
+        generateName: 'openid-',
```
I'd give this a more specific name.

```suggestion
        generateName: 'openid-client-secret-',
```

jcaianirh commented:
@spadgett Added optional ca and replied to review comments except claims. What do you think of making them mandatory and removing the help text?

spadgett commented:
> @spadgett Added optional ca and replied to review comments except claims. What do you think of making them mandatory and removing the help text?

It doesn't look like you need to supply all of them though. We'll need confirmation from the auth team. Each of these is an array in the API.

@enj We need help understanding this :)

jcaianirh commented:
@spadgett @bmignano updated screenshot.

spadgett commented:
@enj I pushed a commit that allows you to set more than one claim value and extra scopes. We probably need better help text on these fields, but it's a start.

@jcaianirh PTAL

<img width="597" alt="Add Identity Provider: OpenID Connect · OKD 2019-03-26 19-36-54" src="https://user-images.githubusercontent.com/1167259/55040405-91540180-4ffe-11e9-8d98-a36e33bffbf0.png">

spadgett commented:
/lgtm
/hold

@jcaianirh I squashed the commits as well. Feel free to hold cancel if this looks good

jcaianirh commented:
/hold cancel

spadgett commented:
/retest

spadgett commented:
It looks like the storage tests might be fixed now.

/retest

spadgett commented:
/retest

spadgett commented:
/retest

spadgett commented:
/retest

spadgett commented:
/retest

spadgett commented:
/retest
